
# Introduction to the Integrative Genomics Viewer (IGV) 


what is igv
what is it useful for (alignments vs signal track and regions vs SNPs)


owen -  pull previous igv stuff from rnaseq workshop
consider which datasets to use...? 



[IGV User Guide](https://software.broadinstitute.org/software/igv/UserGuide)


loading reference genomes

exploring the gene track (+/- strand)


types of ghenomics data that can be read into igv 
viewing alignments in igv (.bai indexes)
specific types of alignments (WGS vs RNA-seq)
hovering over reads 
gray colors in reads vs colored bases for mapping 
empty reads for bad mapping quality 
paired end reads vs single end reads 

colors for bases different to reference 
insertions and deletions 

reading in custom reference genomes 


loading in signal track data from chip seq (bigwig)
load in bed files 
defining regions of interest 

saving and restoring a session 




what are we not covering - advanced variant evaluation and review in igv 
